### ISSUE TITLE
Strange Accuracy

### ISSUE THREAD

**Issue by zmonoid** [dmlc/mxnet#2652]
The training accuracy is so strange here, for 10 classes it always give below 0.1 accuray (Even by random guess should give 0.1 accuray) what maybe the problem?

```
2016-07-08 16:04:35,022 Node[0] start with arguments Namespace(batch_size=128, clip_gradient=5.0, data_dir='../data', data_shape=299, gpus='1,2,3', kv_store='local', load_epoch=None, log_dir='../logs/', log_file=None, lr=0.01, lr_factor=0.8, lr_factor_epoch=1, mean_file='p072_mean_299.bin', model_prefix=None, network='inception-v3', num_classes=10, num_epochs=30, num_examples=20454, save_model_prefix='../saves/p072_757616846', train_dataset='p072_train_299.rec', val_dataset='p072_val_299.rec')
[16:04:35] src/io/iter_image_recordio.cc:211: ImageRecordIOParser: ../data/p072_train_299.rec, use 4 threads for decoding..
[16:04:35] src/io/./iter_normalize.h:103: Load mean image from ../data/p072_mean_299.bin
[16:04:36] src/io/iter_image_recordio.cc:211: ImageRecordIOParser: ../data/p072_val_299.rec, use 4 threads for decoding..
[16:04:36] src/io/./iter_normalize.h:103: Load mean image from ../data/p072_mean_299.bin
2016-07-08 16:04:37,026 Node[0] Start training with [gpu(1), gpu(2), gpu(3)]
2016-07-08 16:06:09,463 Node[0] Epoch[0] Batch [50]     Speed: 143.87 samples/sec       Train-accuracy=0.014531
2016-07-08 16:06:09,464 Node[0] Epoch[0] Batch [50]     Speed: 143.87 samples/sec       Train-cross-entropy=3.109404
2016-07-08 16:06:55,228 Node[0] Epoch[0] Batch [100]    Speed: 139.85 samples/sec       Train-accuracy=0.026875
2016-07-08 16:06:55,228 Node[0] Epoch[0] Batch [100]    Speed: 139.85 samples/sec       Train-cross-entropy=2.414823
2016-07-08 16:07:41,070 Node[0] Epoch[0] Batch [150]    Speed: 139.61 samples/sec       Train-accuracy=0.082344
2016-07-08 16:07:41,070 Node[0] Epoch[0] Batch [150]    Speed: 139.61 samples/sec       Train-cross-entropy=2.349708
2016-07-08 16:07:49,335 Node[0] Update[160]: Change learning rate to 8.00000e-03
2016-07-08 16:08:02,168 Node[0] Epoch[0] Resetting Data Iterator
2016-07-08 16:08:02,169 Node[0] Epoch[0] Time cost=159.022
2016-07-08 16:08:03,012 Node[0] Saved checkpoint to "../saves/p072_757616846-0001.params"
2016-07-08 16:08:04,290 Node[0] Epoch[0] Validation-accuracy=0.229167
2016-07-08 16:08:04,290 Node[0] Epoch[0] Validation-cross-entropy=2.294902
2016-07-08 16:08:49,557 Node[0] Epoch[1] Batch [50]     Speed: 142.33 samples/sec       Train-accuracy=0.055156
2016-07-08 16:08:49,557 Node[0] Epoch[1] Batch [50]     Speed: 142.33 samples/sec       Train-cross-entropy=2.342106
2016-07-08 16:09:35,492 Node[0] Epoch[1] Batch [100]    Speed: 139.33 samples/sec       Train-accuracy=0.044687
2016-07-08 16:09:35,493 Node[0] Epoch[1] Batch [100]    Speed: 139.33 samples/sec       Train-cross-entropy=2.406293
2016-07-08 16:10:16,836 Node[0] Update[319]: Change learning rate to 6.40000e-03
2016-07-08 16:10:21,430 Node[0] Epoch[1] Batch [150]    Speed: 139.32 samples/sec       Train-accuracy=0.072813
2016-07-08 16:10:21,430 Node[0] Epoch[1] Batch [150]    Speed: 139.32 samples/sec       Train-cross-entropy=2.341382
2016-07-08 16:10:41,649 Node[0] Epoch[1] Resetting Data Iterator
2016-07-08 16:10:41,649 Node[0] Epoch[1] Time cost=157.359
2016-07-08 16:10:42,553 Node[0] Saved checkpoint to "../saves/p072_757616846-0002.params"
2016-07-08 16:10:43,842 Node[0] Epoch[1] Validation-accuracy=0.080729
2016-07-08 16:10:43,843 Node[0] Epoch[1] Validation-cross-entropy=2.310077
2016-07-08 16:11:29,275 Node[0] Epoch[2] Batch [50]     Speed: 141.81 samples/sec       Train-accuracy=0.074844
2016-07-08 16:11:29,275 Node[0] Epoch[2] Batch [50]     Speed: 141.81 samples/sec       Train-cross-entropy=2.324910
2016-07-08 16:12:15,243 Node[0] Epoch[2] Batch [100]    Speed: 139.23 samples/sec       Train-accuracy=0.064844
2016-07-08 16:12:15,243 Node[0] Epoch[2] Batch [100]    Speed: 139.23 samples/sec       Train-cross-entropy=2.332900
2016-07-08 16:12:44,739 Node[0] Update[478]: Change learning rate to 5.12000e-03

```

**Comment by zmonoid** [comment]
Also the training accuracy does not agree with validation accuracy. Validation gives as high as 0.3 accuracy, but training gives always less than 0.1 accuracy, why?

**Comment by zmonoid** [comment]
I  got the reason.
The data is not shuffled.

**Comment by yitang** [comment]
@zmonoid Hi, I just wonder why does shuffle the data has such big impact on training?

**Comment by zmonoid** [comment]
@yitang Suppose you got the training samples like index [1...10] are labeled A,  index [11..20] labeled B, etc.

Your training batch size is 11. For the first batch the gradient moves towards regression on A (index 11 sample can be neglected as it contributes little to the loss in one batch), therefore for the next batch [12 ... 23] the network tends to classify all samples to A (result in the accuracy even lower than random guess), but will also drag the gradient towards the regression on B. Well... you would image what will happen next.

**Comment by yitang** [comment]
@zmonoid  Thank you very much for your explanation! It's really helpful. I got the same problem when run VGG with small batch size (6).
